Add barbs plot

jonseddon opened the pull request:
For a personal project it was useful to be able to plot wind barbs. matplotlib.pyplot.barbs does this easily and so I've added it to iris.plot in exactly the same way that quiver is already in there. I've added a gallery example too. 

From looking at #3398 I think that I need to add a line to test the image produced to `tests/results/imagerepo.json` but all the tests are passing locally currently  and I can't work out the hash value to include.

jonseddon commented:
Now that I've seen how Travis runs the tests on the gallery images I think that I can see what I need to do to get the hash and add the image to `SciTools/test-images-scitools`. I'll do that later this week.

pp-mo commented:
>  I can see what I need to do to get the hash and add the image to SciTools/test-images-scitools

There is some well-hidden advice about this in the Developer Guide.
In fact, we recently made an improved version, which you can find the  in the latest master (unreleased) documentation  : available, as usual, from the "docs: latest" badge on the main Readme  ...
https://scitools-docs.github.io/iris/master/developers_guide/graphics_tests.html#developer-graphics-tests

The last sections should be of some help : Basically, you can use "iris/tests/idiff.py" to calculate a hash, fix up the repodata.json etc.

Re-reading this though, I can see that it could still be improved !
The procedures listed on the page are now correct, I hope, but there are some awkward + fussy details which mean some procedures need to be followed _exactly_ to get graphics testing and idiff to run properly : creating an extra temporary directory; running from the right directory, finding and identifying files which it re-names, etc.

jonseddon commented:
Hi @pp-mo, sorry, this is a home project for the sailing club and so I've only just had a chance to look at this again. Thanks for the new v3 graphics tests instructions, they're a great improvement over the v2.4 ones on the website. SciTools/test-iris-imagehash#33 contains the new hash. All tests are now passing apart from `iris.tests.test_image_json.TestImageFile` which should pass once the PR in the other repository has been merged.

jonseddon commented:
Just giving this a quick bump. Now that SciTools/test-iris-imagehash#33 has been merged then if an admin could trigger the tests then hopefully they'll pass.

trexfeathers commented:
Hi @jonseddon , thanks for keeping an eye on this. Please could you rebase your branch onto the latest Iris master? [travis.yml](https://github.com/SciTools/iris/blob/master/.travis.yml) has recently needed to change to point at the latest commit hash for iris-test-data, which is why a couple of tests are still not passing. Thanks

jonseddon commented:
Thanks @trexfeathers for looking at that. It's been rebased and the tests are passing now.

trexfeathers commented:
OK thanks @jonseddon . We're currently looking in general at the docs and gallery examples so I believe @abooton is going to schedule this for some wider team attention 👍

jonseddon commented:
This is just a bit of a bump to see what's going on. Sadly this PR's not on the Kanban board for Iris 3.0. 

Please let me know if you need the gallery example included with this PR updated for the new style gallery that you've introduced.

bjlittle commented:
Hey @jonseddon, your PR needs a wee bit of love... fancy resolving your conflicts with `master`? Which has moved on quite a bit :smiley:

jonseddon commented:
Tests should pass when #3942 is merged in.

bjlittle commented:
@jonseddon This PR has sadly gone rather stale... do you have capacity to rebase and resurrect it?

jonseddon commented:
@bjlittle Wow team scitools has been busy lately! Rebased and hopefully ready for review.

jonseddon commented:
Needs updating (and new image hashes supplied) as per #4110.

pp-mo commented:
HI @jonseddon 
Sorry for huge delays on this.
I see you have said it's good, but have left it in draft.
Can you confirm that this is this good to go, and if not can we get it finished within the next ~2 weeks ?
We have it earmarked for [3.1.0 release](https://github.com/SciTools/iris/discussions/4072), so it would be really good to get this signed off.

jonseddon commented:
Awaiting merge of https://github.com/SciTools/test-iris-imagehash/pull/49 then will push updated commit.

trexfeathers reviewed (changes requested):
Thanks for the addition @jonseddon, very nice! I've played around a bit and it makes sense. Interesting how insistent Iris is on converting back to an SI unit in your gallery script.

This needs three things aside from my comments:

1. An equivalent to `iris.tests.integration.plot.test_vector_plots.TestQuiver`.
2. _Another_ rebase
3. Align your new code with the latest CI standards. Have a look at the existing tests for how we use `# isort:skip`

jonseddon commented on lib/iris/plot.py:
```diff
@@ -1400,6 +1400,54 @@ def _vector_component_args(x_points, y_points, u_data, *args, **kwargs):
+    Args:
+
+    * u_cube, v_cube : (:class:`~iris.cube.Cube`)
+        u and v vector components.  Must have same shape and units of knot.
```
The barbs are different for units of ms-1, so I've clarified this in c99fa2a by removing the units but mentioning the increments corresponding to each lines, triangles, etc.

jonseddon commented:
@trexfeathers I've rebased (after several attempts) and addressed the other comments. 

I'm a bit stuck on "Align your new code with the latest CI standards". Which file should I do this in? The other gallery examples don't include this.

There's some new image hashes for the vector tests to be merged in SciTools/test-iris-imagehash#50 so tests should fail this time.

trexfeathers commented on lib/iris/plot.py:
```diff
@@ -1400,6 +1400,54 @@ def _vector_component_args(x_points, y_points, u_data, *args, **kwargs):
+        At present, if u_cube and v_cube have geographic coordinates, then they
+        must be in a lat-lon coordinate system, though it may be a rotated one.
+        To transform wind values between coordinate systems, use
+        :func:`iris.analysis.cartography.rotate_vectors`.
```
These notes still won't appear unless you have the double-colons: `.. Note::`.

jonseddon replied:
Fixed in both quiver() and  barbs() by d671c270.

trexfeathers commented on docs/gallery_tests/test_plot_wind_barbs.py:
```diff
@@ -0,0 +1,30 @@
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+
+# Import Iris tests first so that some things can be initialised before
+# importing anything else.
+import iris.tests as tests
```
```suggestion
# Import Iris tests first so that some things can be initialised before
# importing anything else.
import iris.tests as tests  # isort:skip
```

You're right that none of the other gallery tests have this yet, but that's an oversight and we've just got lucky so far. The statement about importing `iris.tests` first will stop being true the moment we add another import that isort thinks should go before it.

trexfeathers commented:
>I'm a bit stuck on "Align your new code with the latest CI standards". Which file should I do this in? The other gallery examples don't include this.

My general point was that rebasing onto code where the CI standards had changed wouldn't have adjusted any new code, so there may have been non-compliances with the latest Black, flake8, or especially isort given this is new. Looks like the pre-commit bot caught the only problem (7a8808dacc63efdba65c2924b5c2df7ac49f4273).

Regarding `# isort:skip`: see https://github.com/SciTools/iris/pull/3710#discussion_r685923537

trexfeathers commented:
As far as I'm concerned this is ready to go. @jonseddon are you OK to take it out of draft?

trexfeathers commented:
Thanks for your patience @jonseddon 🏆

jonseddon commented:
> Thanks for your patience @jonseddon trophy

Thanks to yourself and everyone who's helped!
